services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java

```diff
@@ -1077,8 +1077,9 @@ private int getRefreshExpiration() {
                 clientSessionMaxLifespan = realm.getClientSessionMaxLifespan();
             }
 
+            AuthenticatedClientSessionModel clientSession = clientSessionCtx.getClientSession();
             if (clientSessionMaxLifespan > 0) {
-                int clientSessionMaxExpiration = userSession.getStarted() + clientSessionMaxLifespan;
+                int clientSessionMaxExpiration = clientSession.getTimestamp() + clientSessionMaxLifespan;
```

Comment:
I think this is not correct. The lifespan is from the start of the client session, not the last refresh. Something like the following:

```diff
-                int clientSessionMaxExpiration = userSession.getStarted() + clientSessionMaxLifespan;
+                AuthenticatedClientSessionModel clientSession = userSession.getAuthenticatedClientSessionByClient(client.getId());
+                int clientSessionMaxExpiration = (clientSession.getStarted() == 0? clientSession.getTimestamp() : clientSession.getStarted()) + clientSessionMaxLifespan;
```

Does it make sense to you?

Reply:
And please do the same for the offline `getOfflineExpiration` method. It has the same issue.

Reply:
Yeah. Not sure why I missed that at the time of writing the fix. Sadly, I am not sure when I have time to work on this. I since moved on and would need to setup the whole project again + find a slot in my free time.  Please feel free to change my commit or use my commit in a new PR.

Reply:
Okis, no problem, then I'm going to create a new PR. I'm going to joing this issue with https://github.com/keycloak/keycloak/pull/15070 (they are similar and modifying the same code area). I'll respect both commits.